### ISSUE TITLE
PySyft supports Differential Privacy ?

### ISSUE THREAD

**Issue by pepper-jk** [OpenMined/PySyft#3820]
## Question
What mechanics providing **Differential Privacy (DP)** are implemented and how can you access them?

**EDIT** by @LaRiffle 
- Opacus (formerly pytorch-dp) : [See blogpost](https://blog.openmined.org/pysyft-opacus-federated-learning-with-differential-privacy/)
- [PATE](https://github.com/OpenMined/PySyft/blob/53fc01700186b38d5330332007894f8a631c1c08/test/torch/differential_privacy/test_pate.py)

## Further Information
Does PySyft provide a library for DP besides PATE?
If so where does one find and use them?
If not which DP libraries are compatible with PySyft and can be recommended?
I'm assuming [pytorch/opacus](https://github.com/pytorch/opacus), formerly known as [PyTorch-DP](https://github.com/facebookresearch/pytorch-dp), will be compatible right?

Also is there a DP method that is used by the workers by default? #2770 suggests something like this.

Over all it is not clear to me where the DP tool, that [is being advertised](https://github.com/OpenMined/PySyft#high-level-architecture), is implemented and used.

## Problem with the Issue closing Bot

I'm sorry if this was answered anywhere before. I tried to read up on this in your issues, but all of the DP issues seem to have been closed due to inactivity rather than because they were actually solved.

Here is a list of some examples: #1933, #1824, #1953, #2200, #2770

## Additional Context
Over the past week we've been looking into PySyft for **Federated Learning (FL)** as opposed to Tensorflow Federated (TFF).
Our main issue with TFF was, that it had a fixed implementation for differential privacy, which could not be modified regarding how and where in a **FL hierarchy** it was applied.
After doing the tutorials and looking at some more of the other examples, I've found that **PySyft** does the opposite. It allows the developer to implement and apply **any DP technique** and use it at any point of a hierarchy. PySyft merely allows you to build a hierarchy (or network of any structure for that matter) of PyTorch workers.
Even though this solves the issue we had with TFF, it would still burden us to implement the DP mechanics ourselves.

PATE seems like a very specific approach and implementation combining a specific FL structure and DP. We are looking for more general DP functions.

**Comment by pepper-jk** [comment]
Thanks @LaRiffle for removing the `stale` label.

**Comment by catdogpandas** [comment]
Yes. I face the same issue. I tried to use https://github.com/tensorflow/privacy for **TFF**, but failed. It seems I have to make a fresh start ? : ) 😭

**Comment by pepper-jk** [comment]
@catdogpandas do you mean Tensorflow Federated? I thought they had their own DP implementation out of the box.
Has been a while since I tried TFF, but it is kind of a mess. Our biggest problem was tensorflow/federated/issues/832 and in regards to DP the location of clipping and noise application was hard coded.

As a result we disregarded it for our research now and switched to pytorch with [pytorch/opacus](https://github.com/pytorch/opacus) for differential privacy. As we only do research experiments we have not yet implemented a real hierarchy with PySyft yet.

Hope this helps you.

**Comment by catdogpandas** [comment]
Thank you. I will try <a href="https://github.com/pytorch/opacus">pytorch/opacus </a> first.

**Comment by gmuraru** [comment]
I think opacus was also integrated in Syft? @LaRiffle

**Comment by LaRiffle** [comment]
Yes we have a first integration of Opacus!
Check out: https://blog.openmined.org/pysyft-opacus-federated-learning-with-differential-privacy/

**Comment by madhavajay** [comment]
@pepper-jk We have some examples here for Opacus and PyDP: https://github.com/OpenMined/PySyft/tree/dev/examples/differential-privacy

Things are in flux a little bit but we have some big stuff coming and better examples specifically for DP.
I will close this issue for now as DP is being actively worked on and when we have more to share we will announce and update the docs and examples. Thanks for your patience.

**Comment by BambooPalace** [comment]
> Yes we have a first integration of Opacus!
> Check out: https://blog.openmined.org/pysyft-opacus-federated-learning-with-differential-privacy/

hi @LaRiffle , i tried to run the code in this blog just recently with pysyft 0.2.9, but there is an error with parameter has no `grad_sample` attribute. 
I have seen similiar issues asked but unsolved #4917  , so could u please share were there versions of Opacus and Pysyft that can run together at the time of the blog?

**Comment by maodong2056** [comment]
> > Yes we have a first integration of Opacus!
> > Check out: https://blog.openmined.org/pysyft-opacus-federated-learning-with-differential-privacy/
> 
> hi @LaRiffle , i tried to run the code in this blog just recently with pysyft 0.2.9, but there is an error with parameter has no `grad_sample` attribute.
> I have seen similiar issues asked but unsolved #4917 , so could u please share were there versions of Opacus and Pysyft that can run together at the time of the blog?

Have u known versions of Opacus and Pysyft that can run together at this blog
